### PR TITLE
support ms domain

### PR DESCRIPTION
**Pull request by nbcsm**
- extract common constants
avoid to use raw string everywhere.

- enhance extra_opset support for graph
preserve extra_opset information when converting between graph and graph_proto.
add test accordingly.

- support ms domain via extra_opset
support extra_opset via python api and command line.
converter will load onnx domain mapping first, then load mapping for each known domain by order, and apply custom mapping functions (if there is any).
for a given TF op mapping, the later mapping overrides earlier's. e.g. for TF Range op, MS mapping will override onnx mapping.

- support extra opset test
introduce parameterized package to simplify test case parameterization.
refactor test code to test extra opset

- add ms range-1 op support
add ms range-1, add range test for <ms, 1> opset.

i have manually verified extra_opset via command line locally.

I will append a commit to this PR to update readme.md after this PR is reviewed.

### CODE REVIEW

File: tf2onnx/tfonnx.py [hunk 1]
```diff
@@ -2331,13 +2322,28 @@ def tensorflow_onnx_mapping(g, continue_on_error, custom_op_handlers):
...
+
+    # load mapping for onnx domain
     for target_opset, op_map in _OPSETS:
         if target_opset <= g.opset:
             ops_mapping.update(op_map)
 
-    # apply custom ops on top of the assembled opset. We can either completment the opset
+    # load mapping for known extra opsets
+    # order matters, later mapping overrides earlier's
+    if g.extra_opset is not None:
```
Comment (guschmue) [inline tf2onnx/tfonnx.py]: can't we use a dict("domain", mapping) so we can support any custom opsets?

File: examples/custom_op_via_python.py [hunk 2]
```diff
@@ -26,7 +25,7 @@ def print_handler(ctx, node, name, args):
     _ = tf.identity(x_, name="output")
     onnx_graph = tf2onnx.tfonnx.process_tf_graph(sess.graph,
                                                  custom_op_handlers={"Print": print_handler},
```
[the remark this answers is not in the document]
Comment (nbcsm) [reply examples/custom_op_via_python.py]: the api itself is not changed.
regarding the ```extra_opset``` argument value, previous ```helper.make_opsetid(_TENSORFLOW_DOMAIN, 1)``` equals to ```constants.DEFAULT_CUSTOM_OP_OPSET``` here actually.

[the remark this answers is not in the document]
Comment (nbcsm) [reply examples/custom_op_via_python.py]: on a second thought, maybe it is better to keep previously way to invoke the api.
user does not really need to know what is in ```tf2onnx.constants```.
i will revert change to this file.

[the remark this answers is not in the document]
Comment (nbcsm) [reply examples/custom_op_via_python.py]: hmm, need to fix ```custom_op_handlers``` due to https://github.com/onnx/tensorflow-onnx/pull/305

File: tests/test_backend.py [hunk 3]
```diff
@@ -905,60 +906,95 @@ def test_sqrt(self):
         _ = tf.identity(x_, name=_TFOUTPUT)
         self._run_test_case([_OUTPUT], {_INPUT: x_val})
 
-    @check_opset_min_version(7, "cast")
-    def test_range_const(self):
+    def _test_range_const(self, extra_opset=None):
```
[the remark this answers is not in the document]
Comment (nbcsm) [reply tests/test_backend.py]: yep, i have been struggling with whether to add a new file too.

the problem of a new file is: test case code is hard to share between onnx domain and ms domain.
take ReverseSequence op in ms domain for example (we will add it soon after this pr), this op is **implicitly** involved in many lstm/gru/rnn related test cases among several scripts.
if we want to test ReverseSequence in ms domain, we will would like to re-use all these tests.
but we don't really want to copy/paste all these test cases to that new file for ms domain.

File: tf2onnx/custom/__init__.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (nbcsm) [reply tf2onnx/custom/__init__.py]: sure.

File: examples/custom_op_via_python.py [hunk 5]
```diff
@@ -14,7 +13,7 @@ def print_handler(ctx, node, name, args):
     # becomes:
     #   T output = Identity(T Input)
     node.type = "Identity"
-    node.domain = _TENSORFLOW_DOMAIN
+    node.domain = constants.DEFAULT_CUSTOM_OP_OPSET.domain
```
[the remark this answers is not in the document]
Comment (nbcsm) [reply examples/custom_op_via_python.py]: ok

File: tf2onnx/tfonnx.py (hunk 1 above)
Comment (nbcsm) [reply tf2onnx/tfonnx.py]: ok.
i was expecting the coming refactoring the we talked about, so i chose a simple way here for now.

File: tests/test_graph.py [hunk 6]
```diff
@@ -348,11 +352,35 @@ def print_handler(ctx, node, name, args):  # pylint: disable=unused-argument
...
             self.assertEqual(
-                'digraph { input1 [op_type=Placeholder shape="[2, 3]"] Print [op_type=Identity] '
-                'output [op_type=Identity] input1:0 -> Print Print:0 -> output }',
+                'digraph { input1 [op_type=Placeholder shape="[2, 3]"] Print [domain="ai.onnx.converters.tensorflow" '
+                'op_type=Identity] output [op_type=Identity] input1:0 -> Print Print:0 -> output }',
                 onnx_to_graphviz(g))
+            self.assertEqual(g.opset, self.config.opset)
+            self.assertEqual(g.extra_opset, [constants.TENSORFLOW_OPSET])
+
+    def test_extra_opset(self):
```
Comment (guschmue) [inline tests/test_graph.py]: this would fail for any runtime != onnxruntime ?

**Comment by nbcsm** [comment]
thanks for the review.

**Comment by nbcsm** [comment]
https://github.com/onnx/tensorflow-onnx/issues/402

**Pull request merged**